### DISCUSSION TITLE
Can you tell me more about conv2d ( ggml_conv_2d )?

### DISCUSSION THREAD

**Discussion by Cyberhan123** [leejet/stable-diffusion.cpp#103] [General]
I am not a professional, and I would like to know whether there is room for optimization of ggml_conv_2d from an industry perspective, because it is too slow at 512x512.

I searched for relevant information，I'll list them：
ggml_conv_2d: https://github.com/ggerganov/ggml/pull/564#issuecomment-1762953580
pytorch: https://github.com/pytorch/pytorch/pull/49582

**Comment by Cyberhan123** [comment]
From a lazy perspective, we can inherit third-party libraries for testing
https://github.com/oneapi-src/oneDNN
https://github.com/amd/ZenDNN

**Reply by Cyberhan123** [reply]
At present, I have found a more feasible reference: https://github.com/Tencent/ncnn/blob/master/src/layer/x86/convolution_x86.cpp#L280

**Reply by FSSRepo** [reply]
@Cyberhan123 Thank you a lot!!, best research!.

> https://github.com/Tencent/ncnn/blob/master/src/layer/x86/convolution_x86.cpp#L280

 I will try to implement this code to ggml_conv2d to test memory usage and speed

**Reply by Cyberhan123** [reply]
According to my research, the implementation process on the CPU needs to be completely customized instead of using the ggml implementation.

**Reply by Cyberhan123** [reply]
In CUDA, we are actually not in a hurry to implement it because it is more complicated.

**Reply by Cyberhan123** [reply]
No matter what, this is also my learning process, and I am willing to spend time studying these details with you. This is also part of our growth.
